src/runtime_src/core/common/usage_metrics.cpp

```diff
@@ -0,0 +1,467 @@
+    if (auto ctx_it = std::find_if(
+        map_it->second.hw_ctx_vec.begin(), map_it->second.hw_ctx_vec.end(), 
+        [hwctx_handle](hw_ctx_metrics& ctx) {return ctx.handle == hwctx_handle;}); ctx_it != map_it->second.hw_ctx_vec.end()) {
+      if (auto kernel_it = std::find_if(
+          ctx_it->kernel_metrics_vec.begin(), ctx_it->kernel_metrics_vec.end(),
+          [name](kernel_metrics& kernel) {return kernel.name == name;}); kernel_it == ctx_it->kernel_metrics_vec.end()) {
+        struct kernel_metrics k;
+        k.name = name;
+        k.num_args = args;
+        ctx_it->kernel_metrics_vec.emplace_back(k);
+      } 
+    }
+  }
+}
+
+void
+usage_metrics_logger::
+log_kernel_run_info(const xrt::kernel_impl* krnl_impl, const xrt::run_impl* run_hdl, bool start, ert_cmd_state state)
+{
+  // collecting time at start of call as next calls will be overhead
+  auto ts_now = std::chrono::high_resolution_clock::now();
+  try {
+    auto kernel =
+        xrt_core::kernel_int::create_kernel_from_implementation(krnl_impl);
+
+    auto hw_ctx = xrt_core::kernel_int::get_hw_ctx(kernel);
+    auto hwctx_handle = static_cast<xrt_core::hwctx_handle*>(hw_ctx);
+
+    auto dev_id = xrt_core::hw_context_int::get_core_device(hw_ctx)->get_device_id();
+    auto name = kernel.get_name();
+
+    // this condition is not required as device will be present
+    if (auto map_it = m_dev_map.find(dev_id); map_it != m_dev_map.end()) {
+      if (auto ctx_it = std::find_if(
+          map_it->second.hw_ctx_vec.begin(), map_it->second.hw_ctx_vec.end(), 
+              [hwctx_handle](hw_ctx_metrics& ctx) {return ctx.handle == hwctx_handle;}); ctx_it != map_it->second.hw_ctx_vec.end()) {
+        if (auto kernel_it = std::find_if(
+            ctx_it->kernel_metrics_vec.begin(), ctx_it->kernel_metrics_vec.end(),
+            [name](kernel_metrics& kernel) {return kernel.name == name;}); kernel_it != ctx_it->kernel_metrics_vec.end()) {
+          if (start) {
+            // record start everytime because previous run may be finished, timeout, aborted or stopped
+            kernel_it->exec_times[run_hdl].start_time = ts_now;
+            kernel_it->exec_times[run_hdl].is_valid = true;
+          }
+          else {
+            // make start time invalid so we can record for next run, add duration to total time and increment total runs
+            if (kernel_it->exec_times[run_hdl].is_valid) {
+              if (state == ERT_CMD_STATE_COMPLETED) {
+                // valid run increament run
+                kernel_it->total_runs++;
+                kernel_it->total_time += std::chrono::duration_cast<std::chrono::microseconds>(ts_now - kernel_it->exec_times[run_hdl].start_time);
+              }
+              // invalidate start time, run may be finished, aborted or timed out
+              kernel_it->exec_times[run_hdl].is_valid = false;
+            }
+          }
+        }
+      }
+    }
+  }
```

Comment:
Much too deeply nested.   Use early exit, and/or use automatic binding to [first, second], and/or refactor into more functions.

Reply:
Sure Soren